api/mappers/engine.py

```diff
@@ -0,0 +1,404 @@
+
+
+def map_mv_option_to_engine(
+    mv_option: "MultivariateFeatureOption",
+) -> MultivariateFeatureOptionModel:
+    return MultivariateFeatureOptionModel(value=mv_option.value, id=mv_option.id)
+
+
+def map_environment_to_engine(
+    environment: "Environment",
+) -> EnvironmentModel:
+    """
+    Maps Core API's `environments.models.Environment` mode instance to the
+    flag_engine environment document.
+    Before building the document, takes care of resolving relationships and
+    feature versions.
+
+    :param Environment environment: the environment to map
+    :rtype EnvironmentModel
+    """
+    project: "Project" = environment.project
+    organisation: "Organisation" = project.organisation
+
+    # Read relationships.
+    project_segments: List["Segment"] = project.segments.all()
+    project_segment_rules_by_segment_id: Dict[
+        int,
+        Iterable["SegmentRule"],
+    ] = {segment.pk: segment.rules.all() for segment in project_segments}
+    project_segment_feature_states_by_segment_id = _get_project_segment_feature_states(
+        project_segments,
+        environment.pk,
+    )
+    environment_feature_states: List["FeatureState"] = [
+        feature_state
+        for feature_state in environment.feature_states.all()
+        if feature_state.feature_segment_id is None
+        and feature_state.identity_id is None
+    ]
+    all_feature_states = (
+        *environment_feature_states,
+        *chain(*project_segment_feature_states_by_segment_id.values()),
+    )
+    multivariate_feature_state_values_by_feature_state_id = {
+        feature_state.pk: feature_state.multivariate_feature_state_values.all()
+        for feature_state in all_feature_states
+    }
+
+    # Read integrations.
+    integration_configs: dict[str, Optional["EnvironmentIntegrationModel"]] = {
+        attr_name: getattr(environment, attr_name, None)
+        for attr_name in (
+            "amplitude_config",
+            "heap_config",
+            "mixpanel_config",
+            "segment_config",
+        )
+    }
+    dynatrace_config: Optional["DynatraceConfiguration"] = getattr(
+        environment, "dynatrace_config", None
+    )
+    webhook_config: Optional["WebhookConfiguration"] = getattr(
+        environment, "webhook_config", None
+    )
+
+    # No reading from ORM past this point!
+
+    # Prepare relationships.
+    organisation_model = OrganisationModel(
+        id=organisation.pk,
+        name=organisation.name,
+        feature_analytics=organisation.feature_analytics,
+        stop_serving_flags=organisation.stop_serving_flags,
+        persist_trait_data=organisation.persist_trait_data,
+    )
+    project_segment_models = [
+        SegmentModel(
+            id=segment.pk,
+            name=segment.name,
+            rules=[
+                map_segment_rule_to_engine(segment_rule)
+                for segment_rule in project_segment_rules_by_segment_id.pop(segment.pk)
```

Comment:
Related to the [above thread](https://github.com/Flagsmith/flagsmith/pull/2290/files/3506c14a0e2272589416669eefd8e021bb025021#r1236843797), yeah - I'm not sure why we don't just do 

```python
map_segment_rule_to_engine(segment_rule) for segment_rule in segment.rules.all()
```

I don't quite see the benefit to having to create a separate dictionary to hold the data?

Reply:
That's the point of the

```
    # No reading from ORM past this point!
```

comment — I really wanted to concentrate all queryset evaluations in one place so they're not scattered across the code.

Reply:
Done.